app/controllers/application_controller.rb

```diff
@@ -1,10 +1,149 @@
 require './config/environment'
+require 'pry'
 
 class ApplicationController < Sinatra::Base
+  
+  register Sinatra::ActiveRecordExtension
 
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "its_a_secret"
   end
 
+  get '/' do
+    erb :index
+  end
+
+  get '/signup' do
+    if !is_logged_in
+      erb :'registrations/signup'
+    else
+      redirect '/tweets'
+    end 
+  end
+
+  post '/signup' do
+    if params[:username] != '' && params[:password] != '' && params[:email] != ''
+      @user = User.create(username: params[:username], password: params[:password], email: params[:email])
+      session[:id] = @user.id
+      redirect '/tweets'
+    else 
+      redirect '/signup'
+    end
+  end
+
+  get '/tweets' do
+    if is_logged_in
+      @user = current_user
+      @tweets = Tweet.all 
+      erb :'tweets/index'
+    else
+      redirect '/login'
+    end
+  end
+
+
+  get '/login' do
+    if !is_logged_in
+      erb :'sessions/login'
+    else
+      redirect '/tweets'
+    end
+  end
+
+  post '/login' do
+    @user = User.find_by(username: params[:username])
+    if @user && @user.authenticate(params[:password])
+      session[:id] = @user.id
+      redirect '/tweets'
+    else
+      redirect '/login'
+    end 
+  end
+
+  get '/tweets/new' do
+    if is_logged_in
+      erb :'tweets/create'
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/tweets' do
+    if !params[:content].empty? 
+      @tweet = Tweet.create(content: params[:content], user_id: current_user.id)
+      redirect "/tweets/#{@tweet.id}"
+    else
+      redirect '/tweets/new'
+    end 
+  end
+
+  get '/tweets/:id' do
+    if is_logged_in
+      @tweet = Tweet.find(params[:id])
+      erb :'tweets/show'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    if is_logged_in
+      @tweet = Tweet.find(params[:id])
+      erb :'tweets/edit'
+    else
+      redirect '/login'
+    end
+  end
+
+  patch '/tweets/:id' do
+    @tweet = Tweet.find(params[:id])
+    if !params[:content].empty?
+      @tweet.content = params[:content]
+      @tweet.save
+      redirect "/tweets/#{@tweet.id}"
+    else
+      redirect "/tweets/#{@tweet.id}/edit"
+    end
+  end
+
+  delete '/:id/delete' do
+    @tweet = Tweet.find(params[:id]) 
+    if is_logged_in && @tweet.user_id == current_user.id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(:id => params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.